File: app/src/main/java/com/dublikunt/nclient/loginapi/LoadTags.java
package com.dublikunt.nclient.loginapi;

import android.util.JsonReader;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.dublikunt.nclient.adapters.TagsAdapter;
import com.dublikunt.nclient.api.components.Tag;
import com.dublikunt.nclient.api.enums.TagType;
import com.dublikunt.nclient.settings.Global;
import com.dublikunt.nclient.settings.Login;
import com.dublikunt.nclient.utility.LogUtility;
import com.dublikunt.nclient.utility.Utility;

import org.jsoup.Jsoup;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

import java.io.IOException;
import java.io.StringReader;
import java.util.Locale;

import okhttp3.Request;
import okhttp3.Response;

public class LoadTags extends Thread {
    @Nullable
    private final TagsAdapter adapter;

    public LoadTags(@Nullable TagsAdapter adapter) {
        this.adapter = adapter;
    }

    @NonNull
    private Elements getScripts(String url) throws IOException {

        Response response = Global.getClient().newCall(new Request.Builder().url(url).build()).execute();
        Elements x = Jsoup.parse(response.body().byteStream(), null, Utility.getBaseUrl()).getElementsByTag("script");
        response.close();
        return x;
    }

    @NonNull
    private String extractArray(@NonNull Element e) throws StringIndexOutOfBoundsException {
        String t = e.toString();
        return t.substring(t.indexOf('['), t.indexOf(';'));
    }

    private void readTags(@NonNull JsonReader reader) throws IOException {
        reader.beginArray();
        while (reader.hasNext()) {
            Tag tt = new Tag(reader);
            if (tt.getType() != TagType.LANGUAGE && tt.getType() != TagType.CATEGORY) {
                Login.addOnlineTag(tt);
                if (adapter != null) adapter.addItem();
            }
        }
    }

    @Override
    public void run() {
        super.run();
        if (Login.getUser() == null) return;
        String url = String.format(Locale.US, Utility.getBaseUrl() + "users/%s/%s/blacklist",
            Login.getUser().getId(), Login.getUser().getCodename()
        );
        LogUtility.d(url);
        try {
            Elements scripts = getScripts(url);
            analyzeScripts(scripts);
        } catch (IOException | StringIndexOutOfBoundsException e) {
            e.printStackTrace();
        }

    }

    private void analyzeScripts(@NonNull Elements scripts) throws IOException, StringIndexOutOfBoundsException {
        if (scripts.size() > 0) {
            Login.clearOnlineTags();
            String array = Utility.unescapeUnicodeString(extractArray(scripts.last()));
            JsonReader reader = new JsonReader(new StringReader(array));
            readTags(reader);
            reader.close();
        }
    }
}

File: app/src/main/kotlin/com/dublikunt/nclient/loginapi/LoadTags.kt
package com.dublikunt.nclient.loginapi

import android.util.JsonReader
import com.dublikunt.nclient.adapters.TagsAdapter
import com.dublikunt.nclient.api.components.Tag
import com.dublikunt.nclient.enums.TagType
import com.dublikunt.nclient.settings.Global
import com.dublikunt.nclient.settings.Login
import com.dublikunt.nclient.utility.LogUtility.download
import com.dublikunt.nclient.utility.Utility
import com.dublikunt.nclient.utility.Utility.unescapeUnicodeString
import okhttp3.Request
import org.jsoup.Jsoup
import org.jsoup.nodes.Element
import org.jsoup.select.Elements
import java.io.IOException
import java.io.StringReader
import java.util.*

class LoadTags(private val adapter: TagsAdapter?) : Thread() {
    @Throws(IOException::class)
    private fun getScripts(url: String): Elements {
        val response = Global.client!!
            .newCall(Request.Builder().url(url).build()).execute()
        val x = Jsoup.parse(response.body.byteStream(), null, Utility.baseUrl)
            .getElementsByTag("script")
        response.close()
        return x
    }

    @Throws(StringIndexOutOfBoundsException::class)
    private fun extractArray(e: Element?): String {
        val t = e.toString()
        return t.substring(t.indexOf('['), t.indexOf(';'))
    }

    @Throws(IOException::class)
    private fun readTags(reader: JsonReader) {
        reader.beginArray()
        while (reader.hasNext()) {
            val tt = Tag(reader)
            if (tt.type !== TagType.LANGUAGE && tt.type !== TagType.CATEGORY) {
                Login.addOnlineTag(tt)
                adapter?.addItem()
            }
        }
    }

    override fun run() {
        super.run()
        if (Login.user == null) return
        val url = String.format(
            Locale.US, Utility.baseUrl + "users/%s/%s/blacklist",
            Login.user!!.id, Login.user!!.codename
        )
        download(url)
        try {
            val scripts = getScripts(url)
            analyzeScripts(scripts)
        } catch (e: IOException) {
            e.printStackTrace()
        } catch (e: StringIndexOutOfBoundsException) {
            e.printStackTrace()
        }
    }

    @Throws(IOException::class, StringIndexOutOfBoundsException::class)
    private fun analyzeScripts(scripts: Elements) {
        if (scripts.size > 0) {
            Login.clearOnlineTags()
            val array = unescapeUnicodeString(extractArray(scripts.last()))
            val reader = JsonReader(StringReader(array))
            readTags(reader)
            reader.close()
        }
    }
}
